test/jdk/java/net/Socket/CtorFailTest.java

```diff
@@ -0,0 +1,214 @@
+
+        private final AtomicInteger closeInvocationCounter = new AtomicInteger(0);
+
+        private final Exception bindException;
+
+        private final Exception connectException;
+
+        private MockSocketImpl(Exception bindException, Exception connectException) {
+            this.bindException = bindException;
+            this.connectException = connectException;
+        }
+
+        @Override
+        protected void create(boolean stream) {
+            // Do nothing
+        }
+
+        @Override
+        protected void bind(InetAddress host, int port) throws IOException {
+            throwIfPresent(bindException);
+        }
+
+        @Override
+        protected void connect(SocketAddress address, int timeoutMillis) throws IOException {
+            throwIfPresent(connectException);
+        }
+
+        private void throwIfPresent(Exception exception) throws IOException {
+            if (exception != null) {
+                switch (exception) {
+                    case IOException error -> throw error;
+                    case RuntimeException error -> throw error;
+                    default -> throw new IllegalStateException(
+                            "unknown exception type: " + exception.getClass().getCanonicalName());
+                }
+            }
+        }
+
+        // Rest of the `SocketImpl` methods should not be used, hence overriding them to throw `UOE`
+
+        @Override
+        protected void close() {
+            closeInvocationCounter.incrementAndGet();
+        }
+
+        @Override
+        protected void connect(String host, int port) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        protected void connect(InetAddress address, int port) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        protected void listen(int backlog) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        protected void accept(SocketImpl impl) {
+            throw new UnsupportedOperationException();
+
+        }
+
+        @Override
+        protected InputStream getInputStream() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        protected OutputStream getOutputStream() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        protected int available() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        protected void sendUrgentData(int data) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void setOption(int optID, Object value) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Object getOption(int optID) {
+            throw new UnsupportedOperationException();
+        }
+
```

Comment:
IIRC Junit should print the test method arguments before running a test. I'd suggest adding a toString here, so that when a test fails we know which scenario is failing.